staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,616 @@
+// preserve the priority level then the filter can use this to
+// cancel the handler registration.
+func (qs *queueSetImpl) Quiesce(eh fq.EmptyHandler) {
+	qs.lock.Lock()
+	defer qs.lock.Unlock()
+	if eh == nil {
+		qs.emptyHandler = eh
+		return
+	}
+	// Here we check whether there are any requests queued or executing and
+	// if not then fork an invocation of the EmptyHandler.
+	if qs.numRequestsEnqueued == 0 && qs.getRequestsExecuting() == 0 {
+		// fork an invocation of the EmptyHandler.
+		qs.counter.Add(1)
+		go func() {
+			eh.HandleEmpty()
+		}()
+	}
+
+	qs.emptyHandler = eh
+}
+
+// Wait uses the given hashValue as the source of entropy
+// as it shuffle-shards a request into a queue and waits for
+// a decision on what to do with that request.  If tryAnother==true
+//  at return then the QueueSet has become undesirable and the client
+// should try to find a different QueueSet to use; execute and
+// afterExecution are irrelevant in this case.  Otherwise, if execute
+//  then the client should execute  the request and then call
+// afterExecution().  Otherwise the client should not execute the
+//  request and afterExecution is irrelevant.
+func (qs *queueSetImpl) Wait(ctx context.Context, hashValue uint64) (tryAnother, execute bool, afterExecution func()) {
+	// TODO(aaron-prindle) verify locking is correct
+
+	qs.lockAndSyncTime()
+	// TODO(aaron-prindle) verify and test quiescent
+	// A call to Wait while the system is quiescent will be rebuffed by
+	// returning `tryAnother=true`.
+	if qs.emptyHandler != nil {
+		qs.lock.Unlock()
+		return true, false, func() {}
+	}
+
+	// ========================================================================
+	// Step 1:
+	// 1) Start with shuffle sharding, to pick a queue.
+	// 2) Reject old requests that have been waiting too long
+	// 3) Reject current request if there is not enough concurrency shares and
+	// we are at max queue length
+	// 4) If not rejected, create a packet and enqueue
+	req := qs.timeoutOldRequestsAndRejectOrEnqueue(hashValue)
+	// req == nil means that the request was rejected - no remaining
+	// concurrency shares and at max queue length already
+	if req == nil {
+		qs.lock.Unlock()
+		metrics.AddReject(qs.name, "queue-full")
+		return false, false, func() {}
+	}
+	// ========================================================================
+
+	// ------------------------------------------------------------------------
+	// Step 2:
+	// 1) The next step is to invoke the method that dequeues as much as possible.
+
+	// This method runs a loop, as long as there
+	// are non-empty queues and the number currently executing is less than the
+	// assured concurrency value.  The body of the loop uses the fair queuing
+	// technique to pick a queue, dequeue the request at the head of that
+	// queue, increment the count of the number executing, and send true to
+	// the request's channel.
+	qs.dequeueWithChannelAsMuchAsPossible()
```

Comment:
I don't understand why this is done here?

Suppose N requests come in nearly at once in a burst, but no more; they are most or all rate limited. What causes the next dequeue? Before you say the afterCompletion function, let's assume they are all canceled via the context?

Reply:
I guess the purpose of this is that you don't have to run an extra goroutine per queueset that dequeues requests? Why do we think that would be problematic? We already run ~3 or more goroutines *per request*, a standing one per queueset shouldn't be problematic.

When I have this problem, my first approach is usually a goroutine for dequeueing, and the function for this wait does something like:

```
proceed := make(chan struct{})
self.requestChan <- request{id: /*some id*/, doneChan: proceed}

select {
case <-proceed:
  /* do actual work */
  self.requestChan <- request{id: /*some id*/, finished: true}
case <-ctx.Done():
  // cancel the request
  self.requestChan <- request{id: /*some id*/, canceled: true}
}
```

Can you help me understand why we did it where each incoming request's goroutine does a little bit of the work? It seems less understandable and not better in terms of locking, and harder to test. What am I missing?

Reply:
As the code is arranged right now, I'm not sure how you could write a test that would make me feel like it's impossible for the system to get into a condition where there are requests queued but no goroutine is working on dequeueing them. It becomes a thing that needs to be tested about the system as a whole, not just locally, since any caller failing to call afterCompletion can cause a big problem, if I'm reading this right.

Reply:
(1) Regarding cancelation, it can have either of two effects: either (a) it causes the request to be ejected from a queue, or (b) it (attempts to) interrupt execution.  For (a) this is no cause to dispatch new requests.  For (b), I recommend updating the comments on Wait to explicitly say what is implemented in the feature branch: afterExecute is called once _either_ the request finishes executing _or_ the request is canceled.  So afterExecute will do the dispatches in that case.

(2) We should have the invariant that the number of requests dispatched at this point is zero or one.  The current code does not have that property, due to an oversight that I made a comment on.  Once that is fixed, the code will have that property.  Calling `dequeueWithChannelAsMuchAsPossible` will dispatch the right number of requests, whether that is zero, one, or more.  We need a function that general.  I see no need to write a separate function that is less general.

(3) Since the queuing and accounting of requests executing is done in this package, and every change already involves a call into this package, we do not need to fork any extra goroutines to make things happen.  Yes, if the client fails to call afterExecution then things will go wrong.  Yes, that is a property of more than this package.